jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java

```diff
@@ -1037,22 +1038,32 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
                                 LOG.warn("Encountered multiple `Host` headers.  Previous `Host` header already seen as `{}`, new `Host` header has appeared as `{}`", _parsedHost, _valueString);
                             checkViolation(DUPLICATE_HOST_HEADERS);
                         }
+                        if (!MISMATCHED_AUTHORITY.isAllowedBy(_complianceMode))
+                        {
+                            HttpURI httpURI = HttpURI.build().uri(_method == null ? null : _method.toString(), _uri.toString());
```

Comment:
We really don't want to go to the effort here build a `HttpURI` here, only to throw it away and then in fact we have already started building the exact same `HttpURI` in `org.eclipse.jetty.server.HttpChannelOverHttp#startRequest`.

Option A) is to change the RequestHandler signature to 
```java
    public interface RequestHandler extends HttpHandler
    {
        /**
         * This is the method called by parser when the HTTP request line is parsed
         *
         * @param method The method
         * @param uri The raw bytes of the URI.
         * @param version the http version in use
         * @deprecated use {@link #startRequest(String, HttpURI, HttpVersion)}
         */
        @Deprecated
        default void startRequest(String method, String uri, HttpVersion version)
        {}
        
        /**
         * This is the method called by parser when the HTTP request line is parsed
         *
         * @param method The method
         * @param uri The raw bytes of the URI.
         * @param version the http version in use
         */
        default void startRequest(String method, HttpURI uri, HttpVersion version)
        {
            startRequest(method, uri.toString(), version);
        }
    }
```
We can then have a recycled HttpURI.Mutable in the parser and build the URI there, so it is available.

Option B) is to move this check to `HttpChannelOverHttp`

I think option A is OK.  Let me do some experiments and I may have a diff for you....

Reply:
@joakime I have created branch `jetty-10-9312-HttpParser-HttpURI` that has this updated signature.
The only issue is that the field is a Mutable, and we take it as an Immutable as we call startRequest... and that immutable is not available when you need to do this check.   So perhaps we need yet another field to remember the immutable?